[13.0][IMP] stock_reception_screen: minor improvements

sebalix opened the pull request:
- invert the `set_location` and `select_packaging` steps
- filter the list of available destinations based on the package storage type
- product info: display the qty available, outgoing qty and the most recent expiry date

Ref. 1967

- Do not use anymore a compute+inverse field to handle the destination location of the processed move line to avoid sync issues. Instead a dedicated field is now used to handle the user's input (and its initial/default value)

Ref. 2039

sebalix committed:
[IMP] stock_reception screen: invert  steps

Invert the 'set_location' and 'select_packaging' steps.

sebalix committed:
[IMP] filter the list of available destinations

sebalix committed:
[IMP] display more product info

Display the qty available, the outgoing qty and the most recent expiry
date of related lots on the reception screen.

simahawk commented on stock_reception_screen/models/stock_move_line.py:
```diff
@@ -10,6 +10,11 @@ class StockMoveLine(models.Model):
     lot_life_date = fields.Datetime(
         string="End of Life Date", related="lot_id.life_date"
     )
+    original_location_dest_id = fields.Many2one(
+        comodel_name="stock.location",
+        ondelete="restrict",
+        string="Original default destination",
+    )
```
Do you really need this field on move line? It seems used only for this screen.
Can't you have a mapping in for the reception screen only?
See below.

sebalix replied:
After some checks, I rewrote the implementation of this, field removed.

simahawk commented on stock_reception_screen/models/stock_reception_screen.py:
```diff
@@ -124,8 +125,21 @@ class StockReceptionScreen(models.Model):
+    )
     # current move line
     current_move_line_id = fields.Many2one(comodel_name="stock.move.line", copy=False)
+    current_move_line_original_location_dest_id = fields.Many2one(
+        related="current_move_line_id.original_location_dest_id"
+    )
```
```suggestion
    current_move_line_original_location_mapping = Serialized(default={})
    current_move_line_original_location_dest_id = fields.Many2one(
        compute="_compute_current_move_line_original_location_dest_id"
    )
```
In the compute lookup for the mapping that you build on create instead of storing the value on the move line itself.
What do you think?

sebalix replied:
I prefered to store the initial destination location to always propose it to the user (see the domain on the view) even if there are other places where to store the goods, because I suppose the user still don't know if the other places are free, so better to still propose the default value.
And it was store on create because when processing a line on the screen, we could change the destination, then cancel the screen. At this stage we have lost the original value computed by Odoo.

But by writing it maybe you're right: like the package, perhaps we should wait the validation of all the steps to make these data persistant on the move line... it'll require more work to do so.

sebalix replied:
I changed the implementation and dropped this field.

simahawk commented on stock_reception_screen/models/stock_reception_screen.py:
```diff
@@ -246,6 +284,28 @@ def _inverse_current_move_line_package(self):
+    @api.depends("package_storage_type_id.location_storage_type_ids")
+    def _compute_allowed_location_dest_ids(self):
+        for wiz in self:
+            self.allowed_location_dest_ids = self.env["stock.location"].search(
```
```suggestion
            wiz.allowed_location_dest_ids = self.env["stock.location"].search(
```

sebalix replied:
fixed

Commit:
Filter allowed locations on child locations only

Even when we apply the storage type limitations, we have to search on
child locations.
Fix a reference to "self" instead of "wiz" in the initial domain.

Commit:
Keep previously selected dest. location on screen

In the reception screen, when we select a destination location
and click on Next, the location becomes empty and a message asks
to set a dest. location.

This is because on the button click, the inverse method of
current_move_line_location_dest_id writes on move_line.location_dest_id,
and the value of current_move_line_location_dest_id is read again from
the compute method. The compute method erases the field's value.
If a destination has been set previously on the field, keep it instead.
To know if a destination has been set, we check if it is different from
the move, as the default location equal to the move's one.

Additionally, we don't empty the field's value if the allowed locations
is equal to the selected location: this should cover the hypothetical
case where a stock.move would have a leaf location as destination, we
want to keep the same in the move line.

jgrandguillaume reviewed (approved):
LGTM Thx !

sebalix commented:
Issues have been found, do not merge this one please (ref. 2039)

sebalix committed:
Store line destination in a dedicated field

Do not use anymore a compute+inverse field to handle the destination
location of the processed move line to avoid sync issues.
Instead a dedicated field is now used to handle the user's input (and
its initial/default value), and its value is stored as destination of
the move line when the step is validated.

simahawk commented:
> Issues have been found, do not merge this one please (ref. 2039)

@sebalix still the case?

guewen commented:
> still the case?

AFAIS it's fixed by the last commit

sebalix commented:
@simahawk sorry, yes fixed by the last commit

simahawk commented:
@sebalix squash pls :)

sebalix commented:
@simahawk done

guewen commented:
/ocabot merge minor
